README.md
## Synopsis

**Fanap's POD** Chat service

# Recent Changelog

All notable changes to this project will be documented here.

## [Unreleased]

-   Search in threads metadata

In order to see complete list of changelog please visit [ChangeLog](https://github.com/masoudmanson/pod-chat/blob/master/changelog.md)

## Code Example

Create an Javascript file e.x `index.js` and put following code there:

```javascript
var Chat = require('podchat-browser');

var params = {
  appId: new Date().getTime(),
  grantDeviceIdFromSSO: false,
  enableCache: true, // Enable Client side caching
  mapApiKey: "API_KEY_OF_NESHAN_MAP", //  {**REQUIRED**} API Key of Neshan Map
  socketAddress: "ws://172.16.106.26:8003/ws", // {**REQUIRED**} Socket Address
  ssoHost: "http://172.16.110.76", // {**REQUIRED**} Socket Address
  platformHost: "http://172.16.106.26:8080/hamsam", // {**REQUIRED**} Platform Core Address
  fileServer: "http://172.16.106.26:8080/hamsam", // {**REQUIRED**} File Server Address
  serverName: "chat-server", // {**REQUIRED**} Server to to register on
  token: "SSO_TOKEN", // {**REQUIRED**} SSO Token
  wsConnectionWaitTime: 500, // Time out to wait for socket to get ready after open
  connectionRetryInterval: 5000, // Time interval to retry registering device or registering server
  connectionCheckTimeout: 10000, // Socket connection live time on server
  messageTtl: 24 * 60 * 60, // Message time to live (1 day in seonds)
  reconnectOnClose: true, // auto connect to socket after socket close
  asyncLogging: {
    onFunction: true, // log main actions on console
    onMessageReceive: true, // log received messages on console
    onMessageSend: true, // log sent messaged on console
    actualTiming: true // log actual functions running time
  },
  clientName: 'SDK Default example', // used for logging (Sentry)
};

var chatAgent = new Chat(params);
```

## Event Listeners

Listen on these events to get updated data on your client

### chatReady

This is the main Event which fires when your SDK has connected to ASYNC server and gets ready to chat.
**Write your code in chatReady's callback function**

```javascript
chatAgent.on("chatReady", function() {
  /**
   * Your code goes here
   */
});
```

### error

You can get all the Errors here

```javascript
/**
* Listen to Error Messages
*/
chatAgent.on("error", function(error) {
  console.log("Error: ", error.code, error.message);
});
```

### messageEvents

You'll get all the Message Events here

```javascript
/**
 * Listen to Message Events Listener
 */
chatAgent.on("messageEvents", function(event) {
  var type = event.type,
    message = event.result.message;

    switch (type) {
      case "MESSAGE_NEW":
        /**
         * Sending Message Seen to Sender after 5 secs
         */
        setTimeout(function() {
          chatAgent.seen({messageId: message.id, ownerId: message.ownerId});
        }, 5000);

        break;

      case "MESSAGE_EDIT":
        break;

      case "MESSAGE_DELIVERY":
        break;

      case "MESSAGE_SEEN":
        break;

      default:
        break;
    }
});
```

### threadEvents

You'll get all the Events which are related to Threads in threadEvents listener

```javascript
/**
 * Listen to Thread Events
 */
chatAgent.on("threadEvents", function(event) {
  var type = event.type;

  switch (type) {
    case "THREAD_LAST_ACTIVITY_TIME":
      break;

    case "THREAD_NEW":
      break;

    case "THREAD_ADD_PARTICIPANTS":
      break;

    case "THREAD_REMOVE_PARTICIPANTS":
      break;

    case "THREAD_LEAVE_PARTICIPANT":
      break;

    case "THREAD_RENAME":
      break;

    case "THREAD_MUTE":
      break;

    case "THREAD_UNMUTE":
      break;

    case "THREAD_INFO_UPDATED":
      break;

    case "THREAD_UNREAD_COUNT_UPDATED":
      break;

    default:
      break;
  }
});
```

### fileUploadEvents

You'll get all the Events which are related to File Uploads in fileUploadEvents listener

```javascript
/**
 * Listen to File Uploads Events
 */
chatAgent.on("fileUploadEvents", function(event) {
  console.log(event);
});
```

### fileDownloadEvents

You'll get all the Events which are related to File Download Events in fileDownloadEvents listener

```javascript
/**
 * Listen to File Download Events
 */
chatAgent.on("fileDownloadEvents", function(event) {
  console.log(event);
});
```

## User Functions

### getUserInfo

Returns current user's Profile Information

```javascript
chatAgent.getUserInfo(function(userInfo) {
  console.log(userInfo);
});
```

### getNotSeenDuration
Takes an Array of user Ids and Returns the time which those users were away in miliseconds

```javascript
/**
 * Get Unseen Duration of users
 * 
 * @param   {list}  usersIds    An array of userIds
 */
chatAgent.getNotSeenDuration({
    userIds: [122, 123]
}, function(result) {
    console.log(result);
});
```

## Thread Functions

### createThread

```javascript
var createThreadTypes = {
  NORMAL: 0,
  OWNER_GROUP: 1,
  PUBLIC_GROUP: 2,
  CHANNEL_GROUP: 4,
  CHANNEL: 8
}

chatAgent.createThread({
    title: "Thread Title Sample",
    type: "NORMAL",
    invitees: [],
    image: "http://yoursite.com/image.jpg",
    description: "Thread Description",
    metadata: {
      key: value
    }
  }, function(createThreadResult) {
    console.log(createThreadResult);
  }
);
```

### getThreads

```javascript
chatAgent.getThreads({
    count: 50,
    offset: 0,
    name: "A String to search in thread titles and return result",
    threadIds: [] // Array of threadIds to get
  }, function(threadsResult) {
    var threads = threadsResult.result.threads;
    console.log(threads);
  }
);
```

### getHistory

To get messages list of a thread, you can use getHistory() function. Accepted parameters are listed below:

-   threadId {number} ** Id of thread to get it's history ** `required`
-   count {number} ** number of messages to get ** `default is 50`
-   offset {number} ** offset of get query ** `default is 0`
-   order {string} ** order of get query ** `default is DESC`
-   id {string} ** ID of single message to get it's content **
-   uniqueIds {array} ** Array of uniqueIds to get those messages from server **
-   query {string} ** A string to search in messages content **
-   metadataCriteria {json} ** A json object to use as entry for GraphQL seach on message's metaData **
-   fromTime {number} ** Return messages those time is bigger than this ** `13 digits (like 1547972096473)`
-   fromTimeNanos {number} ** Nano Second precision for fromTime ** `9 digits (like 473533000)`
-   fromTimeFull {number} ** Instead of fromTme and fromTimeNanos you can send fromTimeFull ** `19 digits (like 1547972096473533000)`
-   toTime {number} ** Return messages those time is smaller than this ** `13 digits (like 1547972096473)`
-   toTimeNanos {number} ** Nano Second precision for toTime ** `9 digits (like 473533000)`
-   toTimeFull {number} ** Instead of toTime and toTimeNanos you can send toTimeFull ** `19 digits (like 1547972096473533000)`

```javascript
chatAgent.getHistory({
    count: 50,
    offset: 0,
    threadId: threadId
  }, function(historyResult) {
    var history = historyResult.result.history;
    console.log(history);
  }
);
```

### clearHistory
This functions clears history of a given thread.

```javascript
/**
 * CLEAR THREAD HISTORY
 * @param threadId
 */
chatAgent.clearHistory({
    threadId: 1431
}, function(result) {
    console.log("Clear history result", result);
});
```

### getThreadParticipants

```javascript
chatAgent.getThreadParticipants({
    count: 50,
    offset: 0,
    threadId: threadId
  }, function(participantsResult) {
    var participants = participantsResult.result.participants;
    console.log(participants);
  }
);
```

### getThreadAdmins
Returns a list of thread admins

```javascript
chatAgent.getThreadAdmins({
    count: 50,
    offset: 0,
    threadId: threadId
  }, function(adminsResult) {
    var admins = adminsResult.result.participants;
    console.log(admins);
  }
);
```

### setAdmin
You can make an user an admin for a thread who is already a participant of that thread. In order to change a participant into
an admin, all you have to do is to call `setAdmin()` and set parameters as needed.
One important point is that if you set `roleOperation` parameter as `add`, the user will grant the roles you gave, and if
you set `roleOperation` parameter as `remove`, those roles will be ungranted.

```javascript
chatAgent.setAdmin({
    threadId: 10349,
    admins: [
        {
            userId: 123,
            roleOperation: 'add',
            roles: [
                'post_channel_message',
                'edit_message_of_others',
                'delete_message_of_others',
                'add_new_user',
                'remove_user',
                'thread_admin',
                'add_rule_to_user',
                'remove_role_from_user',
                'read_thread',
                'edit_thread'
            ]
        }]
}, function(result) {
    console.log(result);
});
```

These roles are currently available to be set or unset:

- post_channel_message
- edit_message_of_others
- delete_message_of_others
- add_new_user
- remove_user
- thread_admin
- add_rule_to_user
- remove_role_from_user
- read_thread
- edit_thread

### addParticipants

```javascript
chatAgent.addParticipants({
  threadId: threadId,
  contacts: [contactId1, contactId2, ...]
}, function(result) {
  console.log(result);
});
```

### removeParticipants

```javascript
chatAgent.removeParticipants({
  threadId: threadId,
  participants: [participantId1, participantId2, ...]
}, function(result) {
  console.log(result);
});
```

### leaveThread

```javascript
chatAgent.leaveThread({
  threadId: threadId
}, function(result) {
  console.log(result);
});
```

### muteThread

```javascript
chatAgent.muteThread({
    subjectId: threadId
  }, function(result) {
    console.log("Threaded has been successfully muted!");
  }
);
```

### unMuteThread

```javascript
chatAgent.unMuteThread({
    subjectId: threadId
  }, function(result) {
    console.log("Threaded has been successfully unMuted!");
  }
);
```

### renameThread

```javascript
chatAgent.renameThread({
    title: newName,
    threadId: threadId
  }, function(result) {
    console.log("Threaded has been successfully Renamed!");
  }
);
```

### updateThreadInfo

```javascript
chatAgent.updateThreadInfo({
  threadId: threadId,
  image: imageUrl,
  description: "This is a sample description for a thread",
  title: "New title for thread",
  metadata: {
    name: "John Doe"
  }
}, function(result) {
  console.log(result);
});
```

### spamPvThread

If one who is not in your contacts list, creates a P2P thread including you. You can simply SPAM him/her by calling spamPvThread() and giving it the thread's ID. Notice that the thread must be a P2P thread and the thread owner should not be in your contacts list.

```javascript
chatAgent.spamPvThread({
  threadId: P2PThreadId
}, function(result) {
  console.log(result);
});
```

## Contact Functions

### getContacts

```javascript
chatAgent.getContacts({
    count: 50,
    offset: 0
  }, function(contactsResult) {
  var contacts = contactsResult.result.contacts;
  console.log(contacts);
});
```

### getBlocked

This function return a list of people who you have blocked already.

```javascript
chatAgent.getBlocked({
    count: 50,
    offset: 0
  }, function(contactsResult) {
    if (!result.hasError) {
      console.log(result);
    }
});
```

### block

In order to block a contact of yours, you can simply call block() function and give that contact's Id as a parameter.

```javascript
chatAgent.block({
  contactId: 2247,
  // threadId: 1018,
  // userId: 121
}, function(result) {
  if (!result.hasError)
    console.log("Contact has been successfully Blocked!");
});
```

### unblock

To unblock an already blocked contact, call unblock() function with block Id of that blocked contact.

```javascript
chatAgent.unblock({
  blockId: 425,
  // contactId: 2247,
  // threadId: 1018,
  // userId: 122
}, function(result) {
  if (!result.hasError)
    console.log("Contact has been successfully unBlocked!");
});
```

### addContacts

```javascript
chatAgent.addContacts({
  firstName: "Firstname",
  lastName: "Lastname",
  cellphoneNumber: "0999999999",
  email: "mail@gmail.com"
}, function(result) {
  console.log(result);
});
```

### updateContacts

```javascript
chatAgent.updateContacts({
    id: 66, //contact's id
    firstName: "Firstname", // new firstname
    lastName: "Lastname",// new lastname
    cellphoneNumber: "0999999999", // new cellphone number
    email: "mail@gmail.com" //new email
}, function(result) {
  console.log(result);
});
```

### removeContacts

```javascript
chatAgent.removeContacts({
  id: 234 // contact's id
}, function(result) {
  console.log(result);
});
```

### searchContacts

To search in contacts list, you can use searchContacts() function. Accepted parameters to search are listed below:

-   cellphoneNumber {string}
-   email {string}
-   firstName {string}
-   lastName {string}
-   uniqueId {string}
-   id {string}
-   typeCode {string}
-   q {string}
-   offset {number}
-   size {number}
    extra information can be found here [listContacts() Documentation](http://sandbox.pod.land:8080/apidocs/swagger-ui.html?srv=/nzh/listContacts)

```javascript
chatAgent.searchContacts({
  id: 234 // contact's id
}, function(result) {
  console.log(result);
});
```

## Message Functions

### sendTextMessage

```javascript
chatAgent.sendTextMessage({
    threadId: threadId,
    content: messageText
  }, {
  onSent: function(result) {
    console.log("\nYour message has been Sent!\n");
    console.log(result);
  },
  onDeliver: function(result) {
    console.log("\nYour message has been Delivered!\n");
    console.log(result);
  },
  onSeen: function(result) {
    console.log("\nYour message has been Seen!\n");
    console.log(result);
  }
});
```

### resendMessage

```javascript
chatAgent.resendMessage(uniqueId, {
  onSent: function(result) {
    console.log("\nYour message has been Sent!\n");
    console.log(result);
  },
  onDeliver: function(result) {
    console.log("\nYour message has been Delivered!\n");
    console.log(result);
  },
  onSeen: function(result) {
    console.log("\nYour message has been Seen!\n");
    console.log(result);
  }
}); // unique id of message to be resent
```

### cancelMessage

```javascript
chatAgent.cancelMessage(uniqueId); // unique id of message to be canceled
```

### getMessageDeliveredList

```javascript
chatAgent.getMessageDeliveredList({
  messageId: 19623
});
```

### getMessageSeenList

```javascript
chatAgent.getMessageSeenList({
  messageId: 19623
});
```

### editMessage

```javascript
chatAgent.editMessage({
    messageId: messageId,
    content: newMessage
  }, {
  onSent: function(result) {
    console.log("Edited Message has been Sent!");
    console.log(result);
  },
  onDeliver: function(result) {
    console.log("Edited Message has been Delivered!");
    console.log(result);
  },
  onSeen: function(result) {
    console.log("Edited Message has been Seen!");
    console.log(result);
  }
});
```

### deleteMessage

In order to delete a message for all, set `deleteForAll` parameter as `TRUE`.

```javascript
/**
 * DELETE MESSAGE IN THREAD
 * @param {int}      messageId
 * @param {boolean}  deleteForAll
 */
chatAgent.deleteMessage({
  messageId: messageId,
  deleteForAll: false
}, function(result) {
  console.log(result);
});
```

### deleteMultipleMessages

In order to delete an array of messages for all, set `deleteForAll` parameter as `TRUE`.

```javascript
/**
 * DELETE MESSAGE IN THREAD
 * 
 * @param {int}      messageId
 * @param {boolean}  deleteForAll
 */
chatAgent.deleteMultipleMessages({
    threadId: 10298,
    messageIds: [47710, 47709, 47708],
    deleteForAll: true
}, function(result) {
    console.log("Delete Multiple Message Result", result);
});
```



### replyMessage

```javascript
chatAgent.replyMessage({
    threadId: threadId,
    repliedTo: messageId,
    content: message
  }, {
  onSent: function(result) {
    console.log("Your reply message has been Sent!");
    console.log(result);
  },
  onDeliver: function(result) {
    console.log("Your reply message has been Delivered!");
    console.log(result);
  },
  onSeen: function(result) {
    console.log("Your reply message has been Seen!");
    console.log(result);
  }
});
```

### replyFileMessage

```javascript
chatAgent.replyFileMessage({
    threadId: threadId,
    repliedTo: messageId,
    file: file,
    content: message
  }, {
  onSent: function(result) {
    console.log("Your reply message has been Sent!");
    console.log(result);
  },
  onDeliver: function(result) {
    console.log("Your reply message has been Delivered!");
    console.log(result);
  },
  onSeen: function(result) {
    console.log("Your reply message has been Seen!");
    console.log(result);
  }
});
```

### forwardMessage

```javascript
var messagesIds = [2539, 2538, 2537];

chatAgent.forwardMessage({
    subjectId: threadId,
    content: JSON.stringify(messagesIds)
  }, {
  onSent: function(result) {
    console.log(result.uniqueId + " \t has been Sent! (FORWARD)");
  },
  onDeliver: function(result) {
    console.log(result.uniqueId + " \t has been Delivered! (FORWARD)");
  },
  onSeen: function(result) {
    console.log(result.uniqueId + " \t has been Seen! (FORWARD)");
  }
});
```

## File functions

### sendFileMessage

```html
<form>
  <fieldset>
    <legend>Send File Message</legend>
    <input type="file" name="sendFileInput" id="sendFileInput">
    <br>
    <label for="sendFileDescription">Description: </label>
    <input type="text" name="sendFileDescription" id="sendFileDescription">
    <button type="button" name="button" id="sendFileMessage">Send</button>
  </fieldset>
</form>
```

```javascript
document.getElementById("sendFileMessage").addEventListener("click", function() {
  var fileInput = document.getElementById("sendFileInput"),
    image = fileInput.files[0],
    content = document.getElementById("sendFileDescription").value;

  chatAgent.sendFileMessage({
    threadId: 293,
    file: image,
    content: content,
    metaData: {custom_name: "John Doe"}
  }, {
    onSent: function(result) {
      console.log(result.uniqueId + " \t has been Sent!");
    },
    onDeliver: function(result) {
      console.log(result.uniqueId + " \t has been Delivered!");
    },
    onSeen: function(result) {
      console.log(result.uniqueId + " \t has been Seen!");
    }
  });
});
```

### cancelFileUpload

If you want to cancel sending of the message with a file upload, You can get uploading file's uniqueId
immediately after calling `sendFileMessage()` and send it as a parameter to `cancelFileUpload()`

```javascript
  var instantResult = chatAgent.sendFileMessage({
    threadId: 293,
    file: image,
    content: content
  }, {
    onSent: function(result) {},
    onDeliver: function(result) {},
    onSeen: function(result) {}
  });

  chatAgent.cancelFileUpload({
    uniqueId: instantResult.content.file.uniqueId
  }, function() {
    console.log("File Upload has been Canceled!");
  });
```

### uploadFile

```html
<form>
  <fieldset>
    <legend>Upload File</legend>
    <input type="file" name="file" id="fileInput" value="">
    <button type="button" name="button" id="uploadFile">Upload File</button>
    <br>
    <div id="uploadedFile"></div>
  </fieldset>
</form>
```

```javascript
document.getElementById("uploadFile").addEventListener("click", function() {
  var fileInput = document.getElementById("fileInput"),
    file = fileInput.files[0];

  chatAgent.uploadFile({
    file: file,
    fileName: "Test Name"
  }, function(result) {
    if (!result.hasError) {
      var file = result.result;
      document.getElementById("uploadedFile").innerHTML = "<pre><br>Uploaded File Id: " + file.id + "<br>Uploaded File Name : " + file.name + "<br>Uploaded File HashCode : " + file.hashCode + "</pre>";
    }
  });
});
```

### uploadImage

```html
<form>
  <fieldset>
    <legend>Upload Image</legend>
    <input type="file" name="image" id="imageInput" value="">
    <button type="button" name="button" id="uploadImage">Upload Image</button>
    <br>
    <img id="uploadedImage" />
    <div id="uploadedImageData"></div>
  </fieldset>
</form>
```

```javascript
document.getElementById("uploadImage").addEventListener("click", function() {
  var imageInput = document.getElementById("imageInput"),
    image = imageInput.files[0];

  chatAgent.uploadImage({
    image: image,
    fileName: "Test Name",
    xC: 0,
    yC: 0,
    hC: 800,
    wC: 800
  }, function(result) {
    if (!result.hasError) {
      var image = result.result;

changelog.md
# Changelog

All notable changes to this project will be documented here.
to see complete list of changelog please visit [ChangeLog](http://gp.fanapsoft.ir/pod/pod-basic-sdk/chat-js-browser-sdk/-/blob/master/changelog.md)

## [Unreleased]

## [12.9.7] 25-11-2023

### Added
- Method replyPrivately
- Method replyFileMessagePrivately
- Method getThreadsLight
- Method getPinMessages
- Method lastMessageInfo
- Method addReaction
- Method removeReaction
- Method replaceReaction
- Method getReactionsSummaries
- cache for getReactionsSummaries
- Method getReactionsList
- cache for getReactionsList
- Ability to switch async protocols
- uniqueId for ping messages
- Parameter usernames to method removeParticipants
- xss check for messages contents
- requestBlocking for some call actions
- callId to REJECT_GROUP_CALL

### Fixed
- socketState 0 on event chatState was not being fired
- mute/unmute in calls causing performance bug in browsers
- screenShare cropped when sharing a non fullscreen window

### Changed
- Internal changes making the sdk more modular
- Rewrite call module

## [12.9.6] 29-07-2023
## [12.9.6-snapshot.4] 19-07-2023
## [12.9.6-snapshot.3] 19-07-2023
## [12.9.6-snapshot.2] 19-07-2023

### Added
- Implemented mode: auto switch between async protocols

### Fixed
- window is undefined error in Next.js projects
- Modified 31:LAST_SEEN_UPDATED unreadCount cache condition to prevent errors


## [12.9.6-snapshot.1] 11-07-2023

### Changed
- Update podasync-ws-only@2.9.0-snapshot.10


## [12.9.6-snapshot.0] 09-07-2023

### Changed
- Merge default count=25, offset=0
- Merge sdkParams in separate module

###Fixed
- Fixed store error for not loaded thread on type 31 (unreadCount cache)

## [12.9.5] 2023-07-09

### Changed
- remove old cache
- Modification on unreadCount cache to prevent unwanted higher unreadCount on event unread_count_updated

## [12.9.4] 2023-07-02

### Added
- Properties: metadata, systemMetadata to pinMessageVO in result of method getThread
- sample for deleteMultipleMessages
- Fix lastSeenMessageTime (19 Numbers) on type 31:LAST_SEEN_UPDATED
- timeNanos & Changed time (19 Numbers) in pinMessage

### Changed
- Changed cached unreadCount calculation on new messages

## [12.9.3] 2023-06-20
## [12.9.2] 2023-06-18 - deprecated

### Fixed
- Send threadId in event DELETE_THREAD


## [12.9.1] 2023-06-17

### Fixed
- Incorrect version number in runtime

## [12.9.0] 2023-06-12

### Added
- Properties: metadata, systemMetadata to pinMessageVO in result of method getThread

## [12.8.4] 2023-06-17

### Fixed
- Incorrect version number in runtime

## [12.8.3] 2023-06-12 - deprecated

### Added
- UnreadCount cache

## [12.7.2] 2023-06-12

### Added
- prevent request when socket not connected
- Event: THREAD_UNREAD_COUNT_UPDATED on chat: LAST_MESSAGE_DELETED
- Implemented new poor connection event for audio and video
- Method resetCallStream
- Event CALL_RECORDING_FAILED
- uniqueId to socket data
- Params: username, cellphoneNumber to method: getThreadParticipants
- Param: coreUserId to method: getContacts
- Param: onlyNewMessages to method: getHistory
- Params: isGroup, type to method: getThreads
- Prevent ringing in other devices of the call owner

### Changed
- Method shouldNotProcessChatMessage modified
- Method recallThreadParticipant modified
- Merge audioLevel improvements
- Update package async-ws-only
- Method exportChat modified due to the chat server changes
- Modified ping functionality with chat server
- Improved functionality of field: ownerId
- Removed unnecessary call to getAllThreads on asyncReady
- Modified process of call request handling functionality
- Merge: setToken improvement, sdk version

### Fixed
- Incorrect callId in some scenarios
- Bug in CALL_DIVS event in groupCall
- Bug, call participants voice was playing in other devices of the user

## [12.7.2-snapshot.30]
- Updated package podasync-ws-only to version 2.9.0-snapshot.4 to fix retryStep

## [12.7.2-snapshot.29]
- Rollback get user on each socket reconnect to previous functionality
- Updated package podasync-ws-only to fix retryStep

## [12.7.2-snapshot.28]
- Changes to not prevent get user info requests when chat is not ready (but async is ready) Which causes faster get user on reconnects

## [12.7.2-snapshot.27]
- Changed functionality to get user on each socket reconnect (Temporarily)
- Fixed functionality of method enableParticipantsVideoReceive (Was fixed in other versions and merged here)

## [12.7.2-snapshot.26]
- Added additional check to make sure SDK does well even if chat server sends participant instead of participantVO in new messages or history

## [12.7.2-snapshot.24]
- Fixed screenShare problem
- Consider the call as started when SDK receives 74:START_CALL event instead of waiting for response of CREATE_SESSION


## [12.7.2-snapshot.0] - 2022-11-01
### Added
- Added possibility to accept incoming call while user is in an existing call
- Check to prevent request when socket not connected

## [12.7.1] - 2022-10-29
### Added
- fire Event: THREAD_LAST_SEEN_UPDATED on chat message: LAST_SEEN_UPDATED, type: 31
- Retry requests to kurento
- Property: chatId to call requests
- Property: callId to event type: CALL_PARTICIPANT_LEFT
- Grant permissions internally in methods startCall, startGroupCall, acceptCall
- SDK parameter: typeCodeOwnerId
- Object: deviceManager
- Method: deviceManager.grantUserMediaDevicesPermissions
- Event type: SWITCH_TO_GROUP_CALL
- Possibility to know if we are the call starter
- Reset SDK internal data when starting a new call
- Event type: CALL_RECORDING_STARTED
- Property: callId to Event type: CALL_STARTED
- Parameters: fromTime, toTime in method: getThreads
- Method: sendCallSticker
- Method: getCustomerInfo
- Event: CONTACT_THREAD_UPDATE
- Methods: archiveThread/unArchiveThread
- Property: threadId to event: RECEIVE_CALL
- Method: recallThreadParticipant
- Property: seenByAnyAssistant to thread item
- Property: callCenter to userInfo

### Changed

src/lib/constants.js
const chatMessageVOTypes = {
    CREATE_THREAD: 1,
    MESSAGE: 2,
    SENT: 3,
    DELIVERY: 4,
    SEEN: 5,
    PING: 6,
    BLOCK: 7,
    UNBLOCK: 8,
    LEAVE_THREAD: 9,
    ADD_PARTICIPANT: 11,
    GET_STATUS: 12,
    GET_CONTACTS: 13,
    GET_THREADS: 14,
    GET_HISTORY: 15,
    CHANGE_TYPE: 16,
    REMOVED_FROM_THREAD: 17,
    REMOVE_PARTICIPANT: 18,
    MUTE_THREAD: 19,
    UNMUTE_THREAD: 20,
    UPDATE_THREAD_INFO: 21,
    FORWARD_MESSAGE: 22,
    USER_INFO: 23,
    USER_STATUS: 24,
    GET_BLOCKED: 25,
    RELATION_INFO: 26,
    THREAD_PARTICIPANTS: 27,
    EDIT_MESSAGE: 28,
    DELETE_MESSAGE: 29,
    THREAD_INFO_UPDATED: 30,
    LAST_SEEN_UPDATED: 31,
    GET_MESSAGE_DELIVERY_PARTICIPANTS: 32,
    GET_MESSAGE_SEEN_PARTICIPANTS: 33,
    IS_NAME_AVAILABLE: 34,
    JOIN_THREAD: 39,
    BOT_MESSAGE: 40,
    SPAM_PV_THREAD: 41,
    SET_ROLE_TO_USER: 42,
    REMOVE_ROLE_FROM_USER: 43,
    CLEAR_HISTORY: 44,
    SYSTEM_MESSAGE: 46,
    GET_NOT_SEEN_DURATION: 47,
    PIN_THREAD: 48,
    UNPIN_THREAD: 49,
    PIN_MESSAGE: 50,
    UNPIN_MESSAGE: 51,
    UPDATE_CHAT_PROFILE: 52,
    CHANGE_THREAD_PRIVACY: 53,
    GET_PARTICIPANT_ROLES: 54,
    GET_REPORT_REASONS: 56,
    REPORT_THREAD: 57,
    REPORT_USER: 58,
    REPORT_MESSAGE: 59,
    GET_CONTACT_NOT_SEEN_DURATION: 60,
    ALL_UNREAD_MESSAGE_COUNT: 61,
    CREATE_BOT: 62,
    DEFINE_BOT_COMMAND: 63,
    START_BOT: 64,
    STOP_BOT: 65,
    LAST_MESSAGE_DELETED: 66,
    LAST_MESSAGE_EDITED: 67,
    BOT_COMMANDS: 68,
    THREAD_ALL_BOTS: 69,
    CALL_REQUEST: 70,
    ACCEPT_CALL: 71,
    REJECT_CALL: 72,
    RECEIVE_CALL_REQUEST: 73,
    START_CALL: 74,
    END_CALL_REQUEST: 75,
    END_CALL: 76,
    GET_CALLS: 77,
    RECONNECT: 78,
    CONNECT: 79,
    CONTACT_SYNCED: 90,
    GROUP_CALL_REQUEST: 91,
    LEAVE_CALL: 92,
    ADD_CALL_PARTICIPANT: 93,
    CALL_PARTICIPANT_JOINED: 94,
    REMOVE_CALL_PARTICIPANT: 95,
    TERMINATE_CALL: 96,
    MUTE_CALL_PARTICIPANT: 97,
    UNMUTE_CALL_PARTICIPANT: 98,
    CANCEL_GROUP_CALL: 99,
    LOGOUT: 100,
    LOCATION_PING: 101,
    CLOSE_THREAD: 102,
    REMOVE_BOT_COMMANDS: 104,
    SEARCH: 105,
    CONTINUE_SEARCH: 106,
    REGISTER_ASSISTANT: 107,
    DEACTIVATE_ASSISTANT: 108,
    GET_ASSISTANTS: 109,
    ACTIVE_CALL_PARTICIPANTS: 110,
    CALL_SESSION_CREATED: 111,
    IS_BOT_NAME_AVAILABLE: 112,
    TURN_ON_VIDEO_CALL: 113,
    TURN_OFF_VIDEO_CALL: 114,
    ASSISTANT_HISTORY: 115,
    BLOCK_ASSISTANT: 116,
    UNBLOCK_ASSISTANT: 117,
    BLOCKED_ASSISTANTS: 118,
    RECORD_CALL: 121,
    END_RECORD_CALL: 122,
    START_SCREEN_SHARE: 123,
    END_SCREEN_SHARE: 124,
    DELETE_FROM_CALL_HISTORY: 125,
    DESTINED_RECORD_CALL: 126,
    GET_CALLS_TO_JOIN: 129,
    MUTUAL_GROUPS: 130,
    CREATE_TAG: 140,
    EDIT_TAG: 141,
    DELETE_TAG: 142,
    ADD_TAG_PARTICIPANT: 143,
    REMOVE_TAG_PARTICIPANT: 144,
    GET_TAG_LIST: 145,
    DELETE_MESSAGE_THREAD: 151,
    EXPORT_CHAT: 152,
    ADD_CONTACTS: 200,
    REMOVE_CONTACTS: 201,
    CONTACT_THREAD_UPDATE: 220,
    SWITCH_TO_GROUP_CALL_REQUEST: 221,
    RECORD_CALL_STARTED: 222,
    ARCHIVE_THREAD: 223,
    UNARCHIVE_THREAD: 224,
    CALL_STICKER_SYSTEM_MESSAGE: 225,
    CUSTOMER_INFO: 226,
    RECALL_THREAD_PARTICIPANT: 227,
    INQUIRY_CALL: 228,
    CALL_RECORDING_FAILED: 230,
    LAST_MESSAGE_INFO: 234,
    GET_PIN_MESSAGE: 236,
    GET_THREAD_LIGHT: 237,
    REPLY_PRIVATELY: 238,
    ADD_REACTION: 239,
    REPLACE_REACTION: 240,
    REMOVE_REACTION: 241,
    REACTION_LIST: 242,
    GET_MY_REACTION: 243,
    REACTION_COUNT: 244,
    CUSTOMIZE_REACTION: 245,
    SET_ADMIN_ROLE_TO_USER: 250,
    REMOVE_ADMIN_ROLE_FROM_USER: 251,
    ERROR: 999
};

const inviteeVOidTypes = {
    TO_BE_USER_SSO_ID: 1,
    TO_BE_USER_CONTACT_ID: 2,
    TO_BE_USER_CELLPHONE_NUMBER: 3,
    TO_BE_USER_USERNAME: 4,
    TO_BE_USER_ID: 5,
    TO_BE_CORE_USER_ID: 6
};

const createThreadTypes = {
    NORMAL: 0x0,
    OWNER_GROUP: 0x1,
    PUBLIC_GROUP: 0x2,
    CHANNEL_GROUP: 0x4,
    CHANNEL: 0x8,
    NOTIFICATION_CHANNEL: 0x10,
    PUBLIC_THREAD: 0x20,
    PUBLIC_CHANNEL: 0x40,
    SELF: 0x80
};

const chatMessageTypes = {
    TEXT: '1',
    VOICE: '2',
    PICTURE: '3',
    VIDEO: '4',
    SOUND: '5',
    FILE: '6',
    POD_SPACE_PICTURE: '7',
    POD_SPACE_VIDEO: '8',
    POD_SPACE_SOUND: '9',
    POD_SPACE_VOICE: '10',
    POD_SPACE_FILE: '11',
    LINK: '12',
    END_CALL: '13',
    START_CALL: '14',
    STICKER: '15'
};

const assistantActionTypes = {
    REGISTER: 1,
    ACTIVATE: 2,
    DEACTIVATE: 3,
    BLOCK: 4
};

const systemMessageTypes = {
        IS_TYPING: '1',
        RECORD_VOICE: '2',
        UPLOAD_PICTURE: '3',
        UPLOAD_VIDEO: '4',
        UPLOAD_SOUND: '5',
        UPLOAD_FILE: '6'
};

const imageMimeTypes = [
    'image/bmp',
    'image/png',
    'image/tiff',
    'image/x-icon',
    'image/jpeg',
    'image/webp'
];
const imageExtentions = [
    'bmp',
    'png',
    'tiff',
    'tiff2',
    'ico',
    'jpg',
    'jpeg',
    'webp'
];

const callStickerTypes = {
    RAISE_HAND: 'raise_hand',
    LIKE: 'like',
    DISLIKE: 'dislike',
    CLAP: 'clap',
    HEART: 'heart',
    HAPPY: 'happy',
    ANGRY: 'angry',
    CRY: 'cry',
    POWER: 'power',
    BORED: 'bored'
};
const chatStickerTypes = {
    HI_FIVE: 1,
    LIKE: 2,
    HAPPY: 3,
    CRY: 4
};

const callMetaDataTypes = {
    POORCONNECTION: 1,
    POORCONNECTIONRESOLVED: 2,
    CUSTOMUSERMETADATA: 3,
    SCREENSHAREMETADATA: 4
}

const customizeReactionTypes = {
    deActive: 1,
    active: 2,
    customizeReaction: 3
}

export {
    chatMessageVOTypes,
    inviteeVOidTypes,
    createThreadTypes,
    chatMessageTypes,
    assistantActionTypes,
    systemMessageTypes,
    imageMimeTypes,
    imageExtentions,
    callStickerTypes,
    chatStickerTypes,
    callMetaDataTypes,
    customizeReactionTypes
}
